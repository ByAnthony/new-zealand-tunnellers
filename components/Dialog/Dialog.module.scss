@use "app/variables.module.scss";

.dialog-wrapper.dialog-polyfill {
  .dialog {
    position: fixed;
    top: 50%;
    left: 50%;
    display: block;
    z-index: 2000;
  }

  .dialog:not([data-open="true"]) {
    display: none;
  }
}

.dialog {
  width: 100%;
  max-width: 100%;
  height: 100%;
  max-height: 100%;
  padding: 0;
  border: none;
  background: variables.$primary-dark-color;
  overflow-y: scroll;
  scrollbar-width: none;
  z-index: 2000;
  color: variables.$text-color;
}

.dialog-header-container {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  z-index: 2001;
  background-color: variables.$primary-dark-color;
}

.dialog-header {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 0 variables.$m-margin;
  border-bottom: 1px solid variables.$primary-medium-color;

  button {
    all: unset;
    cursor: pointer;

    @include variables.text-link;
  }
}

.dialog-container {
  padding: variables.$l-margin variables.$m-margin variables.$l-margin
    variables.$m-margin;
}

.dialog-footer-container {
  position: fixed;
  bottom: 0;
  left: 0;
  width: 100%;
  z-index: 2001;
  background-color: variables.$primary-dark-color;
}

.dialog-footer {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: variables.$m-margin;
  border-top: 1px solid variables.$primary-medium-color;

  .total-filters {
    font-weight: 500;
  }
}

.dialog-cancel-button {
  all: unset;
  cursor: pointer;

  @include variables.text-link;
}

.dialog-filter-button {
  padding: variables.$xxs-margin variables.$xs-margin;
  font-size: 1rem;
  color: variables.$secondary-color;
  border-radius: variables.$border-radius;
  border: 1px solid variables.$primary-light-color;
  background-color: variables.$primary-medium-color;
  transition:
    border 0.9s ease-in-out,
    background-color 0.9s ease-in-out;
  cursor: pointer;

  &:hover {
    border: 1px solid variables.$secondary-color;
    background-color: variables.$primary-dark-color;
  }

  @include variables.desktop {
    display: none;
  }
}
