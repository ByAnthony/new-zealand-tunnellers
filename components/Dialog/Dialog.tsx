"use client";

import type { ReactNode } from "react";
import { useRef, useEffect } from "react";

import STYLES from "./Dialog.module.scss";

type Props = {
  id: string;
  children: ReactNode;
  handleResetFilters?: () => void;
  isFooterEnabled?: boolean;
  isOpen: boolean;
  onClose: () => void | null;
  title: string;
  totalFiltered?: number;
  total?: number;
};

type DialogProps = {
  isDialogOpen: boolean;
};

// TODO: this check if the browser support the HTML dialog element. We can remove it once we drop support as a business for Safari 14
const dialogSupported = typeof HTMLDialogElement === "function";

const setPageProperties = ({ isDialogOpen }: DialogProps) => {
  document.body.style.overflowY = isDialogOpen ? "hidden" : "visible";

  if (!dialogSupported) {
    document.body.style.position = isDialogOpen ? "fixed" : "relative";
    document.body.style.width = isDialogOpen ? "100%" : "auto";
  }
};

export function Dialog(props: Props) {
  const {
    children,
    id,
    isFooterEnabled,
    isOpen,
    handleResetFilters,
    onClose,
    title,
    totalFiltered,
    total,
  } = props;

  const ref = useRef<HTMLDialogElement>(null);

  useEffect(() => {
    if (isOpen) {
      // There is a bug on older versions of browser using chromium (chrome, firefox, edge >114) where the dialog got an open attribute even before it is opened.
      // Therefore, when trying to open it, it crashes and log an error mentioning the dialog has already an open attribute.
      ref.current?.removeAttribute("open");
      ref.current?.showModal?.();
    } else {
      ref.current?.close?.();
    }

    setPageProperties({ isDialogOpen: isOpen });
    return () => {
      setPageProperties({ isDialogOpen: false });
    };
  }, [isOpen]);

  return (
    <>
      {isOpen ? (
        <div
          className={
            STYLES[`dialog-wrapper ${dialogSupported ? "" : "dialog-polyfill"}`]
          }
        >
          {!dialogSupported && (
            <div id={`${id}-polyfill"`} data-open={isOpen} />
          )}
          <dialog
            id={id}
            className={STYLES.dialog}
            onClose={onClose}
            data-open={isOpen}
            ref={ref}
          >
            <div className={STYLES["dialog-header-container"]}>
              <div className={STYLES["dialog-header"]}>
                <h2>{title}</h2>
                <button onClick={onClose} className={STYLES["close-button"]}>
                  Done
                </button>
              </div>
            </div>
            <div className={STYLES["dialog-container"]}>{children}</div>
            <div className={STYLES["dialog-footer-container"]}>
              {isFooterEnabled && (
                <div className={STYLES["dialog-footer"]}>
                  <button
                    className={STYLES["dialog-cancel-button"]}
                    onClick={handleResetFilters}
                  >
                    Reset filter
                  </button>
                  <div className={STYLES["total-filters"]}>
                    {totalFiltered}/{total}
                  </div>
                </div>
              )}
            </div>
          </dialog>
        </div>
      ) : null}
    </>
  );
}
