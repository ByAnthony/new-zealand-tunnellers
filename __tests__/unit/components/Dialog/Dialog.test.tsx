import { render, screen, fireEvent } from "@testing-library/react";

import "@testing-library/jest-dom";
import { type Dialog as DialogType } from "@/components/Dialog/Dialog";

jest.mock("react", () => jest.requireActual("react"));

describe("Dialog", () => {
  const defaultProps = {
    id: "test-dialog",
    isOpen: true,
    onClose: jest.fn(),
    title: "Test Dialog",
    totalFiltered: 5,
    total: 10,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe("is supported", () => {
    let Dialog: typeof DialogType;

    beforeEach(() => {
      jest.isolateModules(() => {
        ({ Dialog } = jest.requireActual(
          "../../../../components/Dialog/Dialog",
        ));
      });
    });

    test("renders the dialog with the correct title", () => {
      render(
        <Dialog {...defaultProps}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      expect(screen.getByText("Test Dialog")).toBeInTheDocument();
    });

    test("renders the children content", () => {
      render(
        <Dialog {...defaultProps}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      expect(screen.getByText("Dialog Content")).toBeInTheDocument();
    });

    test("calls onClose when the close button is clicked", () => {
      render(
        <Dialog {...defaultProps}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      const closeButton = screen.getByText("Done");
      fireEvent.click(closeButton);
      expect(defaultProps.onClose).toHaveBeenCalledTimes(1);
    });

    test("renders the footer with the correct totalFiltered and total values", () => {
      render(
        <Dialog {...defaultProps} isFooterEnabled={true}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      expect(screen.getByText("5/10")).toBeInTheDocument();
    });

    test("calls handleResetFilters when the reset button is clicked", () => {
      const handleResetFilters = jest.fn();
      render(
        <Dialog
          {...defaultProps}
          isFooterEnabled={true}
          handleResetFilters={handleResetFilters}
        >
          <div>Dialog Content</div>
        </Dialog>,
      );
      const resetButton = screen.getByText("Reset filter");
      fireEvent.click(resetButton);
      expect(handleResetFilters).toHaveBeenCalledTimes(1);
    });

    test("sets page properties when the dialog is open", () => {
      render(
        <Dialog {...defaultProps}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      expect(document.body.style.overflowY).toBe("hidden");
    });

    test("resets page properties when the dialog is closed", () => {
      render(
        <Dialog {...defaultProps} isOpen={false}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      expect(document.body.style.overflowY).toBe("visible");
    });
  });

  describe("is not supported", () => {
    let htmlDialogElement: typeof window.HTMLDialogElement;

    let Dialog: typeof DialogType;

    beforeEach(async () => {
      htmlDialogElement = window.HTMLDialogElement;
      window.HTMLDialogElement = undefined!;

      jest.isolateModules(() => {
        ({ Dialog } = jest.requireActual(
          "../../../../components/Dialog/Dialog",
        ));
      });
    });

    afterEach(() => {
      window.HTMLDialogElement = htmlDialogElement;
    });

    test("renders the dialog with the correct title", () => {
      render(
        <Dialog {...defaultProps}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      expect(screen.getByText("Test Dialog")).toBeInTheDocument();
    });

    test("renders the children content", () => {
      render(
        <Dialog {...defaultProps}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      expect(screen.getByText("Dialog Content")).toBeInTheDocument();
    });

    test("calls onClose when the close button is clicked", () => {
      render(
        <Dialog {...defaultProps}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      const closeButton = screen.getByText("Done");
      fireEvent.click(closeButton);
      expect(defaultProps.onClose).toHaveBeenCalledTimes(1);
    });

    test("renders the footer with the correct totalFiltered and total values", () => {
      render(
        <Dialog {...defaultProps} isFooterEnabled={true}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      expect(screen.getByText("5/10")).toBeInTheDocument();
    });

    test("calls handleResetFilters when the reset button is clicked", () => {
      const handleResetFilters = jest.fn();
      render(
        <Dialog
          {...defaultProps}
          isFooterEnabled={true}
          handleResetFilters={handleResetFilters}
        >
          <div>Dialog Content</div>
        </Dialog>,
      );
      const resetButton = screen.getByText("Reset filter");
      fireEvent.click(resetButton);
      expect(handleResetFilters).toHaveBeenCalledTimes(1);
    });

    test("sets page properties when the dialog is open", () => {
      render(
        <Dialog {...defaultProps}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      expect(document.body.style.overflowY).toBe("hidden");
      expect(document.body.style.position).toBe("fixed");
      expect(document.body.style.width).toBe("100%");
    });

    test("resets page properties when the dialog is closed", () => {
      render(
        <Dialog {...defaultProps} isOpen={false}>
          <div>Dialog Content</div>
        </Dialog>,
      );
      expect(document.body.style.overflowY).toBe("visible");
      expect(document.body.style.position).toBe("relative");
      expect(document.body.style.width).toBe("auto");
    });
  });
});
