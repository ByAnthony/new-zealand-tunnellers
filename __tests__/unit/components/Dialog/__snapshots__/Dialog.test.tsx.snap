// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog is not supported should render correctly with polyfill and content 1`] = `
<DocumentFragment>
  <div
    class="dialog-wrapper dialog-polyfill"
  >
    <div
      data-open="true"
      id="test-dialog-polyfill"
    />
    <dialog
      class="dialog"
      data-open="true"
      id="test-dialog"
    >
      <div
        class="dialog-header-container"
      >
        <div
          class="dialog-header"
        >
          <h2>
            Test Dialog
          </h2>
          <button
            class="close-button"
          >
            Done
          </button>
        </div>
      </div>
      <div
        class="dialog-container"
      >
        <div>
          Content
        </div>
      </div>
      <div
        class="dialog-footer-container"
      />
    </dialog>
  </div>
</DocumentFragment>
`;

exports[`Dialog is supported should render correctly with content 1`] = `
<DocumentFragment>
  <div
    class="dialog-wrapper "
  >
    <dialog
      class="dialog"
      data-open="true"
      id="test-dialog"
    >
      <div
        class="dialog-header-container"
      >
        <div
          class="dialog-header"
        >
          <h2>
            Test Dialog
          </h2>
          <button
            class="close-button"
          >
            Done
          </button>
        </div>
      </div>
      <div
        class="dialog-container"
      >
        <div>
          Content
        </div>
      </div>
      <div
        class="dialog-footer-container"
      />
    </dialog>
  </div>
</DocumentFragment>
`;
